src/file/SDL_iostream.c

```diff
@@ -731,6 +732,109 @@ SDL_IOStream *SDL_IOFromConstMem(const void *mem, size_t size)
     return iostr;
 }
 
+typedef struct IOStreamDynamicMemData
+{
+    SDL_IOStream *stream;
+    IOStreamMemData data;
+    Uint8 *end;
+} IOStreamDynamicMemData;
+
+static Sint64 SDLCALL dynamic_mem_size(void *userdata)
+{
+    IOStreamDynamicMemData *iodata = (IOStreamDynamicMemData *) userdata;
+    return mem_size(&iodata->data);
+}
+
+static Sint64 SDLCALL dynamic_mem_seek(void *userdata, Sint64 offset, int whence)
+{
+    IOStreamDynamicMemData *iodata = (IOStreamDynamicMemData *) userdata;
+    return mem_seek(&iodata->data, offset, whence);
+}
+
+static size_t SDLCALL dynamic_mem_read(void *userdata, void *ptr, size_t size, SDL_IOStatus *status)
+{
+    IOStreamDynamicMemData *iodata = (IOStreamDynamicMemData *) userdata;
+    return mem_io(&iodata->data, ptr, iodata->data.here, size);
+}
+
+static int dynamic_mem_realloc(IOStreamDynamicMemData *iodata, size_t size)
+{
+    size_t chunksize = (size_t)SDL_GetNumberProperty(SDL_GetIOProperties(iodata->stream), SDL_PROP_IOSTREAM_DYNAMIC_CHUNKSIZE_NUMBER, 0);
+    if (!chunksize) {
+        chunksize = 1024;
+    }
+
+    // We're intentionally allocating more memory than needed so it can be null terminated
+    size_t chunks = (((iodata->end - iodata->data.base) + size) / chunksize) + 1;
+    size_t length = (chunks * chunksize);
+    Uint8 *base = (Uint8 *)SDL_realloc(iodata->data.base, length);
+    if (!base) {
+        return -1;
+    }
+
+    size_t here_offset = (iodata->data.here - iodata->data.base);
+    size_t stop_offset = (iodata->data.stop - iodata->data.base);
+    iodata->data.base = base;
+    iodata->data.here = base + here_offset;
+    iodata->data.stop = base + stop_offset;
+    iodata->end = base + length;
+    return SDL_SetProperty(SDL_GetIOProperties(iodata->stream), SDL_PROP_IOSTREAM_DYNAMIC_MEMORY_POINTER, base);
+}
+
+static size_t SDLCALL dynamic_mem_write(void *userdata, const void *ptr, size_t size, SDL_IOStatus *status)
+{
+    IOStreamDynamicMemData *iodata = (IOStreamDynamicMemData *) userdata;
+    if (size > (iodata->data.stop - iodata->data.here)) {
+        if (size > (iodata->end - iodata->data.here)) {
+            if (dynamic_mem_realloc(iodata, size) < 0) {
+                return 0;
+            }
+        }
+        iodata->data.stop = iodata->data.here + size;
+    }
+    return mem_io(&iodata->data, iodata->data.here, ptr, size);
+}
+
+static int SDLCALL dynamic_mem_close(void *userdata)
+{
+    const IOStreamDynamicMemData *iodata = (IOStreamDynamicMemData *) userdata;
+    void *mem = SDL_GetProperty(SDL_GetIOProperties(iodata->stream), SDL_PROP_IOSTREAM_DYNAMIC_MEMORY_POINTER, NULL);
+    if (mem) {
+        SDL_free(mem);
```

Comment:
No sense in checking if `mem` is NULL; SDL_free will, too.

Reply:
Habit